### PR TITLE
Support markdown headings

### PR DESCRIPTION
**Pull request by dgileadi**
Adds support for Markdown headings. Note that many test HTML files were affected by the accompanying CSS changes.

Per the proposed documentation:

### Headings

A long documentation section can be subdivided by adding headings. Two styles of headings are available.

**ATX Headings:**
An ATX heading is a line of text that starts with one to six `#` characters followed by whitespace and then the heading text. The number of `#` characters determines the heading level. ATX headings may optionally end with any number of trailing `#` characters.
```
/**
 * # H1
 * ## H2
 * ### H3
 * #### H4 ###
 * ##### H5 ##
 * ###### H6 #
 */
```

**Setext Headings:**
A setext heading is one or more lines of text followed by a line of `=` or `*` characters. Those familiar with [Markdown](https://daringfireball.net/projects/markdown/syntax) will notice that Ddoc uses asterisks (`***`) instead of dashes (`---`) for second-level headings to avoid conflicting with [embedded code](https://dlang.org/spec/ddoc.html#embedded_code).
```
/**
 * H1
 * ==
 *
 * A long H2 heading that
 * spans more than one line
 * ************************
 */
```

### CODE REVIEW

**Comment by thewilsonator** [comment]
Thanks! This one is going to take a bit longer to review...

**Comment by dgileadi** [comment]
I suspected it would :)

**Comment by thewilsonator** [comment]
In general please try to align the `=` signs in the ddoc comments.
There are also a few extra instead blank lines that don't need to be there. 

I'm going to put `72h -> merge` on this, but I encourage others to review this as well.

**Comment by dgileadi** [comment]
Should params be aligned like this?
```
Params:
 buf   = an OutBuffer containing the DDoc
 i     = the index within `buf` of the first `#` character
```

**Comment by thewilsonator** [comment]
Yes.

**Comment by WalterBright** [comment]
Why do we need both the ATX and Setext styles? I'd prefer to just stick with ATX.

File: src/dmd/doc.d [hunk 1]
```diff
@@ -2009,14 +2009,183 @@ Lno:
...
+ *                  macro. Upon function return its value is set to `0`.
+ *  i             = an index within `buf`. If `i` is after `iAt` then it gets
+ *                  reduced by the length of the removed macro.
+ */
+private void removeBlankLineMacro(OutBuffer *buf, ref size_t iAt, ref size_t i)
+{
+    if (!iAt)
+        return;
+
+    buf.remove(iAt, 17); // length of "$(DDOC_BLANKLINE)"
```
Comment (WalterBright) [inline src/dmd/doc.d]: You could replace `17` with `$(DDOC_BLANKLINE)".length` and remove the comment.

File: src/dmd/doc.d [hunk 2]
```diff
@@ -2009,14 +2009,183 @@ Lno:
...
+ *          the detected heading level from 1 to 6, or
+ *          0 if not at an ATX heading
+ */
+private int detectAtxHeadingLevel(OutBuffer *buf, size_t i)
+{
+    if (!global.params.markdown)
+        return 0;
+
+    const iHeadingStart = i;
+    size_t iSkipped = skipChars(buf, i, "#");
```
Comment (WalterBright) [inline src/dmd/doc.d]: should be `const` too

File: src/dmd/doc.d [hunk 3]
```diff
@@ -2009,14 +2009,183 @@ Lno:
...
+    const headingLevel = cast(int) (iSkipped - iHeadingStart);
+    if (headingLevel > 6)
+        return 0;
+
+    i = skipChars(buf, iSkipped, " \t");
```
Comment (WalterBright) [inline src/dmd/doc.d]: Don't reuse old variables. Declare a new one. That way the new one can be `const`, too.

File: src/dmd/doc.d [hunk 4]
```diff
@@ -2009,14 +2009,183 @@ Lno:
...
+                iSuffixStart = j;
+            break;
+        case ' ':
+        case '\t':
+            if (!iWhitespaceStart)
+                iWhitespaceStart = j;
+            break;
+        case '\r':
+        case '\n':
+            goto LendHeadingSuffix;
```
Comment (WalterBright) [inline src/dmd/doc.d]: `break;`

File: src/dmd/doc.d [hunk 5]
```diff
@@ -2009,14 +2009,183 @@ Lno:
...
+        default:
+            iSuffixStart = 0;
+            iWhitespaceStart = 0;
```
Comment (WalterBright) [inline src/dmd/doc.d]: `continue;`

File: src/dmd/doc.d [hunk 6]
```diff
@@ -2009,14 +2009,183 @@ Lno:
...
+        }
```
Comment (WalterBright) [inline src/dmd/doc.d]: `break;`

File: src/dmd/doc.d [hunk 7]
```diff
@@ -2009,14 +2009,183 @@ Lno:
...
+    }
+LendHeadingSuffix:
```
Comment (WalterBright) [inline src/dmd/doc.d]: remove label

File: src/dmd/doc.d [hunk 8]
```diff
@@ -2009,14 +2009,183 @@ Lno:
...
+    size_t iWhitespaceStart = j;
+    const slice = buf.peekSlice();
+    for (; j < slice.length; j++)
+    {
+        switch (slice[j])
+        {
+        case '#':
+            if (iWhitespaceStart && !iSuffixStart)
+                iSuffixStart = j;
+            break;
```
Comment (WalterBright) [inline src/dmd/doc.d]: `continue;`

File: src/dmd/doc.d [hunk 9]
```diff
@@ -2009,14 +2009,183 @@ Lno:
...
+        case ' ':
+        case '\t':
+            if (!iWhitespaceStart)
+                iWhitespaceStart = j;
+            break;
```
Comment (WalterBright) [inline src/dmd/doc.d]: `continue;`

File: src/dmd/doc.d [hunk 10]
```diff
@@ -2009,14 +2009,183 @@ Lno:
...
+/****************************************************
+ * Remove any trailing `##` suffix from an ATX-style heading.
+ * Params:
+ *  buf   = an OutBuffer containing the DDoc
+ *  i     = the index within `buf` to start looking for a suffix at
+ */
+private void removeAnyAtxHeadingSuffix(OutBuffer *buf, size_t i)
+{
+    size_t j = i;
+    size_t iSuffixStart = 0;
```
Comment (WalterBright) [inline src/dmd/doc.d]: `const`

**Comment by jacob-carlborg** [comment]
> In general please try to align the = signs in the ddoc comments.

I strongly disagree. If a new item is added it’s most likely that all items need to be realigned and that causes unnecessary diffs.

**Comment by WalterBright** [comment]
@dgileadi well done on writing readable code. BTW, these smaller PRs are clearly easier to review than the all-in-one mega PR.

File: src/dmd/doc.d [hunk 11]
```diff
@@ -2009,14 +2009,183 @@ Lno:
...
+
+/****************************************************
+ * Detect the level of an ATX-style heading, e.g. `## This is a heading` would
+ * have a level of `2`.
+ * Params:
+ *  buf   = an OutBuffer containing the DDoc
+ *  i     = the index within `buf` of the first `#` character
+ * Returns:
+ *          the detected heading level from 1 to 6, or
+ *          0 if not at an ATX heading
```
Comment (jacob-carlborg) [inline src/dmd/doc.d]: I don't think you should aligned cross sections, it will just add unnecessary whitespace forcing this line to be split.

File: src/dmd/doc.d [hunk 12]
```diff
@@ -2009,14 +2009,183 @@ Lno:
...
+    }
+LendHeadingSuffix:
+    if (iSuffixStart)
+        buf.remove(iWhitespaceStart, j - iWhitespaceStart);
+}
+
+/**
+ * Strip the underline from a setext-style heading, e.g. remove all `=`'s and
+ * surrounding whitespace from:
+ * ---
```
Comment (jacob-carlborg) [inline src/dmd/doc.d]: `---` is for highlighting D code.

File: src/dmd/doc.d [hunk 13]
```diff
@@ -2009,14 +2009,183 @@ Lno:
...
+ * This is a heading
+ * =================
+ * ---
+ * Params:
+ *  buf               = an OutBuffer containing the DDoc
+ *  i                 = the index within `buf` of the first underline character
+ *  iLineStart        = the index within `buf` of the start of this line
+ *  iParagraphStart   = the index within `buf` of the most recent paragraph. May
```
Comment (jacob-carlborg) [inline src/dmd/doc.d]: I recommend having just one space after the longest parameter name.

File: test/compilable/ddoc_markdown_headings.d [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+## heading with another trailing literal#
+## heading with backslash-escaped trailing #\##
+
+## Some empty headers:
+##
+#
+### ###
+
+Setext-Style Headings
+=====================
```
Comment (jacob-carlborg) [inline test/compilable/ddoc_markdown_headings.d]: Does it work to have fewer `=` than the number of letters?

**Comment by jacob-carlborg** [comment]
> Why do we need both the ATX and Setext styles? I'd prefer to just stick with ATX.

Either we should support all of the markdown features or we shouldn't call it markdown.

**Comment by jacob-carlborg** [comment]
@dgileadi in the future, please create a separate commit for code style changes (i.e. whitespace, indentation and so on).

File: src/dmd/doc.d (hunk 4 above)
Comment (dgileadi) [reply src/dmd/doc.d]: I believe `break` would exit the switch statement but not the enclosing loop. That's why I used the `goto`.

File: test/compilable/ddoc_markdown_headings.d (hunk 14 above)
Comment (dgileadi) [reply test/compilable/ddoc_markdown_headings.d]: Yes—any number will do per the [CommonMark spec](https://spec.commonmark.org/0.28/#setext-headings). In short, the rules are that the line may only contain whitespace and `=` or `*` characters, and all the `=` or `*` characters must be consecutive. And of course the previous line must contain text.

**Comment by dgileadi** [comment]
> @dgileadi in the future, please create a separate commit for code style changes (i.e. whitespace, indentation and so on).

Will do—sorry for the mess this time.

**Comment by dgileadi** [comment]
> @dgileadi well done on writing readable code. BTW, these smaller PRs are clearly easier to review than the all-in-one mega PR.

Thank you—I'll try to keep the readability going.

**Comment by WalterBright** [comment]
> Either we should support all of the markdown features or we shouldn't call it markdown.

The Setext implementation for Ddoc already diverges from the markdown spec, so this is moot. The divergence is also another reason to dump Setext - there is no compatibility reason to support Setext.

**Comment by dgileadi** [comment]
>> Either we should support all of the markdown features or we shouldn't call it markdown.
>
> The Setext implementation for Ddoc already diverges from the markdown spec, so this is moot. The divergence is also another reason to dump Setext - there is no compatibility reason to support Setext.

While there are some parts of Markdown that I feel very strongly that we should support (like ``` and links, PRs coming soon), I don't feel super strongly about setext headers.

For reference, [adrdox supports the `===` version](http://dpldocs.info/experimental-docs/adrdox.syntax.html#headers) but not `---` or `***`, and [harboured-mod supports them with `===` and `***`](https://github.com/dlang-community/harbored-mod#subheading). So for compatibility with Ddoc alternatives it makes sense to me to keep the support. The `***` syntax does irk me though.

**Comment by jacob-carlborg** [comment]
> The Setext implementation for Ddoc already diverges from the markdown spec, so this is moot. The divergence is also another reason to dump Setext - there is no compatibility reason to support Setext

Then we should stop calling it "markdown".

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
I know this has only been two days but the queue is stuck with the Doc builder and circle doing their own thing so I'm merging this because its green and has been reviewed.

**Comment by WalterBright** [comment]
@thewilsonator pleased don't merge PRs that I have objected to. The Setext needs to go.

**Comment by WalterBright** [comment]
@dgileadi please submit a PR to remove the Setext support.
